crux_core/src/type_generation/facet.rs

```diff
@@ -0,0 +1,516 @@
+    ///
+    /// # Errors
+    /// Returns a [`TypeGenError`] if the type tracing fails.
+    fn register_types(generator: &mut TypeGen) -> Result;
+}
+
+impl Export for () {
+    fn register_types(_generator: &mut TypeGen) -> Result {
+        Ok(())
+    }
+}
+
+/// The `TypeGen` struct stores the registered types so that they can be generated for foreign languages
+/// use `TypeGen::new()` to create an instance
+pub struct TypeGen {
+    pub state: State,
+}
+
+impl Default for TypeGen {
+    fn default() -> Self {
+        TypeGen {
+            state: State::Registering(Registry::default()),
+        }
+    }
+}
+
+impl TypeGen {
+    /// Creates an instance of the `TypeGen` struct
+    #[must_use]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Register all the types used in app `A` to be shared with the Shell.
+    ///
+    /// Do this before calling `TypeGen::swift`, `TypeGen::java` or `TypeGen::typescript`.
+    /// This method would normally be called in a build.rs file of a sister crate responsible for
+    /// creating "foreign language" type definitions for the FFI boundary.
+    /// See the section on
+    /// [creating the shared types crate](https://redbadger.github.io/crux/getting_started/core.html#create-the-shared-types-crate)
+    /// in the Crux book for more information.
+    ///
+    /// # Errors
+    ///
+    /// If any of the types used in the app cannot be registered, an error will be returned.
+    pub fn register_app<'a, A: App>(&mut self) -> Result
+    where
+        A::Effect: Export,
+        A::Event: Deserialize<'static> + Facet<'a>,
+        A::ViewModel: Deserialize<'static> + Facet<'a> + 'static,
+    {
+        A::Effect::register_types(self)?;
+        self.register_type::<A::Event>()?;
+        self.register_type::<A::ViewModel>()?;
+
+        Ok(())
+    }
+
+    /// For each of the types that you want to share with the Shell, call this method:
+    /// e.g.
+    /// ```rust
+    /// # use crux_core::type_generation::facet::TypeGen;
+    /// # use serde::{Serialize, Deserialize};
+    /// # use anyhow::Error;
+    /// #[derive(facet::Facet, Serialize, Deserialize)]
+    /// #[repr(C)]
+    /// enum MyNestedEnum { None }
+    /// #[derive(facet::Facet, Serialize, Deserialize)]
+    /// #[repr(C)]
+    /// enum MyEnum { None, Nested(MyNestedEnum) }
+    /// fn register() -> Result<(), Error> {
+    ///   let mut typegen = TypeGen::new();
+    ///   typegen.register_type::<MyEnum>()?;
+    ///   typegen.register_type::<MyNestedEnum>()?;
+    ///   Ok(())
+    /// }
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// Errors that can occur during type registration.
+    pub fn register_type<'a, 'de, T>(&mut self) -> Result
+    where
+        T: serde::Deserialize<'de> + Facet<'a>,
+    {
+        match &mut self.state {
+            State::Registering(registry) => {
+                let incoming = facet_generate::reflect::<T>();
```

Comment:
Does `facet_generate::reflect` return a serde-generate compatible IR? 

That might be a good, low-risk way of keeping compatibility - have a `reflect_serde` next to `reflect`, where the former returns a serde-generate compatible IR, and `reflect` returns our one with additional info.